### PR TITLE
GH-117881: fix athrow().throw()/asend().throw() concurrent access

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-117881

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2024-04-15-07-37-09.gh-issue-117881.07H0wI.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+prevent concurrect access to an async generator via athrow().throw() or asend().throw()
```
Comment (graingert) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-15-07-37-09.gh-issue-117881.07H0wI.rst]:
```suggestion
prevent concurrent access to an async generator via athrow().throw() or asend().throw()
```

File: Objects/genobject.c [hunk 2]
```diff
@@ -1771,6 +1771,7 @@ async_gen_asend_send(PyAsyncGenASend *o, PyObject *arg)
 
     if (o->ags_state == AWAITABLE_STATE_INIT) {
         if (o->ags_gen->ag_running_async) {
+            o->ags_state = AWAITABLE_STATE_CLOSED;
```
Comment (graingert) [inline Objects/genobject.c:1774]: this was set in `async_gen_athrow_send` but forgotten in `async_gen_asend_send`, and also results in an unawaited coroutine warning

Comment (graingert) [reply Objects/genobject.c:1774]: https://github.com/python/cpython/blob/8418dcc5d3190a199d2236e99e30f2c146939e95/Objects/genobject.c#L2098-L2113

File: Objects/genobject.c [hunk 3]
```diff
@@ -2206,6 +2220,25 @@ async_gen_athrow_throw(PyAsyncGenAThrow *o, PyObject *const *args, Py_ssize_t na
...
+            }
+            return NULL;
+        }
+
+        o->agt_state = AWAITABLE_STATE_ITER;
+    }
+
     retval = gen_throw((PyGenObject*)o->agt_gen, args, nargs);
     if (o->agt_args) {
         return async_gen_unwrap_value(o->agt_gen, retval);
```
Comment (graingert) [inline Objects/genobject.c]:
```suggestion
        retval = async_gen_unwrap_value(o->agt_gen, retval);
        if (retval == NULL) {
            o->agt_gen->ag_running_async = 0;
        }
        return retval;
```

File: Objects/genobject.c [hunk 4]
```diff
@@ -1814,10 +1815,23 @@ async_gen_asend_throw(PyAsyncGenASend *o, PyObject *const *args, Py_ssize_t narg
...
+    if (o->ags_state == AWAITABLE_STATE_INIT) {
+        if (o->ags_gen->ag_running_async) {
+            o->ags_state = AWAITABLE_STATE_CLOSED;
+            PyErr_SetString(
+                PyExc_RuntimeError,
+                "anext(): asynchronous generator is already running");
+            return NULL;
+        }
+
+        o->ags_state = AWAITABLE_STATE_ITER;
```
Comment (graingert) [inline Objects/genobject.c:1827]:
```suggestion
        o->ags_state = AWAITABLE_STATE_ITER;
        o->ags_gen->ag_running_async = 1;
```

File: Objects/genobject.c [hunk 5]
```diff
@@ -2206,9 +2220,32 @@ async_gen_athrow_throw(PyAsyncGenAThrow *o, PyObject *const *args, Py_ssize_t na
...
+            }
+            else {
+                PyErr_SetString(
+                    PyExc_RuntimeError,
+                    "athrow(): asynchronous generator is already running");
+            }
+            return NULL;
+        }
+
+        o->agt_state = AWAITABLE_STATE_ITER;
```
Comment (graingert) [inline Objects/genobject.c:2240]:
```suggestion
        o->agt_state = AWAITABLE_STATE_ITER;
        o->agt_gen->ag_running_async = 1;
```

**Review by encukou — COMMENTED** [review]
As in #117851: I'm not an expert in this area.
It makes sense to start the generator in send/throw.

What I don't get is the purpose of all the synchronization between `ags_state` and `ags_gen->ag_running_async`/`ags_gen->ag_closed`. What's the difference between what they represent? Should they go out of sync at some point?

**Comment by encukou** [comment]
Is it a bug if they go out of sync?
If it is, and `ag_running_async` & `ag_closed` hold the necessary info, why have `ags_state` at all?

**Comment by graingert** [comment]
this is where they go out of sync:

```python
import types
import itertools

@types.coroutine
def _async_yield(v):
    return (yield v)


async def agenfn():
    for i in itertools.count():
        await _async_yield(i)
    return
    yield

agen = agenfn()
gen = agen.__anext__()
print(f"{gen.send(None)=}")
print(f"{agen.ag_running=}")
gen.close()  # uh-oh, gen->ags_state is CLOSED
print(f"{agen.ag_running=}")  # but the async generator is still running
agen.aclose().send(None)  # and it's broken so we can't aclose it
```

**Review by encukou — APPROVED** [review]
The PR looks good to me. But, it would be great to have a more comprehensive way to find issues like this. I commented on the issue.

File: Lib/test/test_asyncgen.py [hunk 6]
```diff
@@ -393,6 +393,131 @@ async def gen():
                 r'anext\(\): asynchronous generator is already running'):
             an.__next__()
 
+    def test_async_gen_asend_throw_concurrent_with_send(self):
```
Comment (hugovk) [inline Lib/test/test_asyncgen.py:400]: [Ignore this review, I'm testing if Bedevere will run without needing approval.]

**Review by hugovk — DISMISSED** [review]

**Comment by hugovk** [comment]
No need for me to review, I was only testing the bot earlier. 👍

File: Objects/genobject.c (hunk 2 above)
Comment (graingert) [reply Objects/genobject.c:1774]: probably could do with a test

Comment (graingert) [reply Objects/genobject.c:1774]: test added in 1220aee3d7a1a8e72903abc815fdf96f31c5ce34

File: Lib/test/test_asyncgen.py [hunk 7]
```diff
@@ -393,6 +393,151 @@ async def gen():
...
+        gen = agen.asend(None)
+        gen.send(None)
+        gen2 = agen.athrow(MyExc)
+
+        with self.assertRaisesRegex(RuntimeError,
+                r'athrow\(\): asynchronous generator is already running'):
+            gen2.throw(MyExc)
+
+        with self.assertRaisesRegex(RuntimeError,
+            r"cannot reuse already awaited aclose\(\)/athrow\(\)"):
```
Comment (graingert) [inline Lib/test/test_asyncgen.py]:
```suggestion
                r"cannot reuse already awaited aclose\(\)/athrow\(\)"):
```

File: Lib/test/test_asyncgen.py [hunk 8]
```diff
@@ -1809,9 +1957,56 @@ class MyException(Exception):
         g = gen()
         with self.assertRaises(MyException):
             g.aclose().throw(MyException)
-            del g
```
Comment (graingert) [inline Lib/test/test_asyncgen.py:1812]: driveby: oops this wasn't running

**Comment by graingert** [comment]
@encukou is there anything else I need to do on this? I'd like to start looking at fixing athrow().close() and asend().close() next

**Comment by encukou** [comment]
No, sorry for the delay!

**Pull request merged**